Update monorepo setup & configs

wottpal opened the pull request:
Overall I'm pretty satisfied with the new setup. Installing dependencies works faster and more reliably via `pnpm` and subsequent `dev`/`build` tasks, too – thanks to `turborepo`. Regarding the configuration files I'm fine with the current state even if I  found the previous way a little bit more declarative. 

* Switched package manager to pnpm
    * **IMPORTANT**: I recommend everyone to delete all its `node_modules` folders, also within the sub-packages, as well as the top-level `package-lock.json`. Then install `pnpm` globally via `npm i -g pnpm` (also see `README.md`) and install dependencies via `pnpm i`. Scripts can be run via `pnpm [script]`.
* Moved prettierrc-configuration to root and let others extend it 
* Moved lint-staged-configuration to top as well as re-configured husky pre-commit script to continue working (it normally assumes the configurations are within the packages but i found a workaround)
* Created a base eslintrc-configuration within the root where the others extend from (it contains the least common shared configuration of both packages)
* Set up turborepo and configured pipelines for `build`, `dev`, and `format` in `turbo.json` (we can do more later).
    * **NOTE**: For you everything should work the same when starting the global dev/build/format script (e.g. via `pnpm dev`. Just turborepo is now used under the hood.
* Fixed some minor issue in `packages/frontend/lib/creatorReadHooks.ts` that should have been dune while updating wagmi but it just came to light (I'm not 100% sure why, I think npm installed some dependencies/versions it should have not. And now pnpm did it correctly. But now it works as it should)

~~Disclaimer: Probably everything breaks when I first try to deploy it to Vercel :D~~

**Update**: Deploy is fixed now and works with the new setup. ❤️

sebastianst commented on packages/frontend/.eslintrc.json:
```diff
@@ -1,13 +1,6 @@
 {
   "root": true,
```
We can remove this and then remove `"../../.eslintrc.json"` from `"extends"`, right? The root `eslintrc` will then be merged automatically.

raulrpearson replied:
I believe so:
https://eslint.org/docs/latest/user-guide/configuring/configuration-files#cascading-and-hierarchy

I also thinks that's usually the way to do it.

sebastianst commented on packages/hardhat/.eslintrc.json:
```diff
@@ -1,6 +1,4 @@
-  "parser": "@typescript-eslint/parser",
-  "plugins": ["@typescript-eslint"],
-  "extends": ["eslint:recommended", "plugin:@typescript-eslint/recommended", "prettier"]
+  "extends": ["../../.eslintrc.json", "prettier"]
```
If we put `"prettier"` into the root `eslintrc`, we can completely remove this file. Note that `"prettier"` is also in `extends` in the `frontend` package.

wottpal replied:
i'm not sure. somewhere i read that `prettier` must be the last item in that array to work properly. but don't remember where. @raulrpearson can you confirm or should i have we have another look?

wottpal replied:
found it. it's written here: https://paulintrognon.fr/blog/typescript-prettier-eslint-next-js. but it's not official documentation.

raulrpearson replied:
yes, the prettier should typically be last because it actually disables a bunch of lint rules (that other configs might have activated) that become irrelevant when prettier is taking care of the formatting.

BTW, I think the similarities between configs will start to melt away once we add linting and formatting of solidity code.

raulrpearson replied:
https://prettier.io/docs/en/integrating-with-linters.html
https://github.com/prettier/eslint-config-prettier#installation

wottpal replied:
yep. i agree with raul. moved eslint files back to package-level in [dc6246d](https://github.com/yieldgate/yieldgate/pull/29/commits/dc6246d77ed27d3878781edcc5b556a4a8a5360f).

sebastianst commented on packages/hardhat/.prettierrc.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  ...require('../../.prettierrc.json'),
+}
```
Looking at the [prettier docs](https://prettier.io/docs/en/configuration.html) I think you can remove this file, prettier will automatically search in parent dirs.

(However, we'd need to reintroduce this if we wanna override/add configuration...)

wottpal replied:
you are right. removed it in [a582afe](https://github.com/yieldgate/yieldgate/pull/29/commits/a582afe1400d71e4338fe0fcace8ab8ee72b9333)

sebastianst commented on packages/frontend/.prettierrc.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  ...require('../../.prettierrc.json'),
+}
```
_See comment on `packages/hardhat/.prettierrc.js`_

wottpal replied:
you are right. removed it in [a582afe](https://github.com/yieldgate/yieldgate/pull/29/commits/a582afe1400d71e4338fe0fcace8ab8ee72b9333)

raulrpearson commented on .husky/pre-commit:
```diff
@@ -4,4 +4,4 @@
 [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
 
-npx lint-staged
+cat .lintstagedrc.json | npx lint-staged --config -
```
How come is this required? Isn't lint-staged picking up the config implicitly when using `npx lint-staged`?

raulrpearson commented on .npmrc:
```diff
@@ -0,0 +1,3 @@
+strict-peer-dependencies = false
```
Being [strict](https://pnpm.io/npmrc#strict-peer-dependencies) about peer dependencies sounds like a good thing. If there's a good reason to set this as false, I think we should at least document it.

raulrpearson commented on .npmrc:
```diff
@@ -0,0 +1,3 @@
+strict-peer-dependencies = false
+auto-install-peers = true
+enable-pre-post-scripts = true
```
I don't have a very strong opinion about this but I like the [arguments](https://github.com/pnpm/pnpm/issues/2891) against setting this option. I'm fine with keeping it as is. Just wanted to share so we can keep it in mind.

wottpal replied:
```
strict-peer-dependencies = false
auto-install-peers = true
```
had some issues with dependencies in the exact same setup (next + hardhat + pnpm) previously that made me to add those lines. but can't recall. i'll try without.

```
enable-pre-post-scripts = true
```
we needed this before turborepo to compile contracts before starting the dev environment. but good catch, now we don't need it anymore.

wottpal replied:
yeah i remembered correctly (see output below). if you're fine i would keep the first two lines, i always liked the behavior of automatically having peers installed more... 

but let me know if you have a strong opinion about it.

```
 ERR_PNPM_PEER_DEP_ISSUES  Unmet peer dependencies

.
├─┬ eslint-config-custom
│ ├─┬ eslint-config-next
│ │ ├── ✕ missing peer eslint@"^7.23.0 || ^8.0.0"
│ │ ├─┬ @typescript-eslint/parser
│ │ │ └── ✕ missing peer eslint@"^6.0.0 || ^7.0.0 || ^8.0.0"
│ │ ├─┬ eslint-import-resolver-typescript
│ │ │ └── ✕ missing peer eslint@"*"
│ │ ├─┬ eslint-plugin-import
│ │ │ └── ✕ missing peer eslint@"^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8"
│ │ ├─┬ eslint-plugin-jsx-a11y
│ │ │ └── ✕ missing peer eslint@"^3 || ^4 || ^5 || ^6 || ^7 || ^8"
│ │ ├─┬ eslint-plugin-react
│ │ │ └── ✕ missing peer eslint@"^3 || ^4 || ^5 || ^6 || ^7 || ^8"
│ │ └─┬ eslint-plugin-react-hooks
│ │   └── ✕ missing peer eslint@"^3.0.0 || ^4.0.0 || ^5.0.0 || ^6.0.0 || ^7.0.0 || ^8.0.0-0"
│ ├─┬ eslint-config-prettier
│ │ └── ✕ missing peer eslint@>=7.0.0
│ └─┬ eslint-plugin-react
│   └── ✕ missing peer eslint@"^3 || ^4 || ^5 || ^6 || ^7 || ^8"
└─┬ next
  ├── ✕ missing peer react@"^17.0.2 || ^18.0.0-0"
  ├── ✕ missing peer react-dom@"^17.0.2 || ^18.0.0-0"
  ├─┬ styled-jsx
  │ └── ✕ missing peer react@">= 16.8.0 || 17.x.x || ^18.0.0-0 || ^17.0.2"
  └─┬ use-sync-external-store
    └── ✕ missing peer react@"^16.8.0 || ^17.0.0 || ^18.0.0 || ^17.0.2"
Peer dependencies that should be installed:
  eslint@">=7.23.0 <8.0.0 || >=8.0.0-0 <8.0.0 || >=8.0.0 <9.0.0"      react-dom@"^17.0.2 || ^18.0.0-0"                                    react@">=17.0.2 <18.0.0 || >=18.0.0-0 <18.0.0 || >=18.0.0 <19.0.0"  

packages/frontend
├─┬ babel-jest
│ ├── ✕ missing peer @babel/core@^7.8.0
│ └─┬ babel-preset-jest
│   ├── ✕ missing peer @babel/core@^7.0.0
│   └─┬ babel-preset-current-node-syntax
│     ├── ✕ missing peer @babel/core@^7.0.0
│     ├─┬ @babel/plugin-syntax-async-generators
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-bigint
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-class-properties
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-import-meta
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-json-strings
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-logical-assignment-operators
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-nullish-coalescing-operator
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-numeric-separator
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-object-rest-spread
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-optional-catch-binding
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     ├─┬ @babel/plugin-syntax-optional-chaining
│     │ └── ✕ missing peer @babel/core@^7.0.0-0
│     └─┬ @babel/plugin-syntax-top-level-await
│       └── ✕ missing peer @babel/core@^7.0.0-0
├─┬ @emotion/react
│ ├── ✕ missing peer @babel/core@^7.0.0
│ └─┬ @emotion/babel-plugin
│   ├── ✕ missing peer @babel/core@^7.0.0
│   └─┬ @babel/plugin-syntax-jsx
│     └── ✕ missing peer @babel/core@^7.0.0-0
├─┬ @emotion/styled
│ └── ✕ missing peer @babel/core@^7.0.0
├─┬ next
│ └─┬ styled-jsx
│   └── ✕ missing peer @babel/core@"*"
└─┬ wagmi
  └─┬ @coinbase/wallet-sdk
    └─┬ eth-block-tracker
      └─┬ @babel/plugin-transform-runtime
        ├── ✕ missing peer @babel/core@^7.0.0-0
        ├─┬ babel-plugin-polyfill-corejs2
        │ ├── ✕ missing peer @babel/core@^7.0.0-0
        │ └─┬ @babel/helper-define-polyfill-provider
        │   ├── ✕ missing peer @babel/core@^7.4.0-0
        │   └─┬ @babel/helper-compilation-targets
        │     └── ✕ missing peer @babel/core@^7.0.0
        ├─┬ babel-plugin-polyfill-corejs3
        │ └── ✕ missing peer @babel/core@^7.0.0-0
        └─┬ babel-plugin-polyfill-regenerator
          └── ✕ missing peer @babel/core@^7.0.0-0
Peer dependencies that should be installed:
  @babel/core@">=7.8.0 <8.0.0"  

packages/hardhat
└─┬ @typechain/ethers-v5
  └── ✕ missing peer @ethersproject/bytes@^5.0.0
Peer dependencies that should be installed:
  @ethersproject/bytes@^5.0.0  

hint: If you want peer dependencies to be automatically installed, add "auto-install-peers=true" to an .npmrc file at the root of your project.
hint: If you don't want pnpm to fail on peer dependency issues, add "strict-peer-dependencies=false" to an .npmrc file at the root of your project.
```

wottpal replied:
update: actually i would like to keep all three lines as esp. the third matches better with the default npm-behavior and somehow also makes more sense imo.

raulrpearson commented on packages/frontend/package.json:
```diff
@@ -7,47 +7,49 @@
-    "format": "prettier --write .",
-    "lint": "eslint --ext .js,.ts,.tsx --fix .",
+    "lint": "eslint --ext .js,.jsx,.ts,.tsx --fix .",
+    "format": "pnpm lint && prettier --write .",
```
I'd be in favour of keeping `format` as a prettier script (no lint).

In turn I'd suggest we create a validate script (or maybe a turbo pipeline?) called validate, which aggregates linting, formatting, type-checking and testing in the appropriate order. Then we know validate is the script we run before committing or pushing.

What do you think?

wottpal replied:
i run those commands manually very rarlely anyways 🤷‍♂️ or do you mean by "is the script we run before committing or pushing" that we configure lint-staged to run something like `pnpm turbo run validate`?

raulrpearson replied:
So my first point is that I don't like running lint inside a script called format. It's not a big deal, it's just that I would have chosen a different name. How about "fix"? Also are we ever going to run lint without wanting to run prettier write afterwards? If so, maybe we could just merge both into a script called fix?

My second point is that I usually like having a "validate" script that runs a bunch of code quality validations: type checking, lint, formatting, tests and build. But I think we can take that discussion out of this PR. I expect it will surface once we start adding tests.

raulrpearson commented on packages/hardhat/package.json:
```diff
@@ -3,13 +3,15 @@
+    "deploy": "pnpm hardhat run scripts/deploy.ts",
+    "deploy:mumbai": "pnpm hardhat run scripts/deploy.ts --network mumbai",
+    "lint": "eslint --ext .js,.jsx,.ts,.tsx --fix .",
+    "format": "pnpm lint && prettier --write ."
```
Same comment that made for the frontend package.

raulrpearson commented:
As I mentioned in our conversations, I think we're ending up with higher complexity without that much deduplication since the shared configs are not that high. Having said that, now that it's been done, I'm happy to keep it and we can always change our mind in the future if it becomes a headache.

BTW, do we want to add Solidity linting and prettifying as I had intended to do in #28?

wottpal commented:
* removed package-level prettier-files (as suggested by seb)
* moved the eslint files back to package-level (as i agree with raul)
* added solidity linting via solhint (there are currently some warnings shown by the linter – see below. but i think this is something contract-related and for a different pr by seb. maybe also the `.solhint.json` needs to be re-configured as the rules are just the default ones – feel free to continue on this.)
*  reverted husky pre-commit as the old version now magically works for me too 🧙‍♂️

feel free to take it over from here, guys. 🇫🇷

sebastianst reviewed (approved):
LGTM

raulrpearson reviewed (approved):
I have slight concerns that we're just sweeping potential dependency problems under the rug with `strict-peer-dependencies = false`.

See [first comment](https://github.com/yieldgate/yieldgate/pull/29#discussion_r923466931), [discussion](https://github.com/yieldgate/yieldgate/pull/29#discussion_r923469758) and [arguments against that setting](https://github.com/pnpm/pnpm/issues/1142#issuecomment-390414329).

I also forgot to ask Dennis why are we setting React 17 as the allowed peer dependency.

But at this point I think it's best to move forward so approving 👍

wottpal commented:
* please go forward regarding the peer-deps setting. as long as the install works, I'm fine ;)
* regarding the react v17 fixture this was related to some react v18 bug i had during/right after the hackathon. i bet it's resolved already as I'm using react v18 in other projects with a similar setup. so we should update it soon :)
